ethereum-network-partition-test/ethereum_helpers.star

```diff
@@ -0,0 +1,88 @@
+# Ethereum API allows us to input "block_number-latest" to return the latest block information. Very helpful!
+LATEST_BLOCK_NUMBER_GENERIC = "latest"
+
+# We have this complex jq filter to remove the `0x` prefix on the hex string returned by the Ethereum node
+# and pad the hexadecimal string to 20 characters (which should be a limit we hopefully never hit)
+# This is a hack to get the hexadecimal block numbers to be comparable between each other
+# We have the equivalent function in Starlark below (see `pad`)
+HEX_PAD_NUMBER = 20
+HEX_PAD_NUMBER_FOR_JQ = HEX_PAD_NUMBER + 2  # +2 to account for the additional `0x` at the beginning of the hex string
+JQ_PAD_HEX_FILTER = """def pad(i): (({0} - (i | length)) * "0") + (i | sub("0x"; "")); pad({1})"""
+
+BLOCK_NUMBER_FIELD = "block-number"
+BLOCK_HASH_FIELD = "block-hash"
+
+
+def get_block(plan, node_id, block_number_hex):
+    """
+    Returns the block information for block number `block_number_hex` (which should be a hexadecimal string starting
+    with `0x`, i.e. `0x2d`)
+
+    The object returned is a struct with 2 fields `number` and `hash`, both hexadecimal encoded strings.
+    """
+    block_response = plan.request(
+        recipe=get_block_recipe(node_id, block_number_hex),
+    )
+    return struct(
+        number=block_response[extracted_field_name(BLOCK_NUMBER_FIELD)],
+        hash=block_response[extracted_field_name(BLOCK_HASH_FIELD)],
+    )
+
+
+def wait_until_node_reached_block(plan, node_id, target_block_number_hex):
+    """
+    This function blocks until the node `node_id` has reached block number `target_block_number_hex`.
+
+    If node has already produced this block, it returns immediately.
+    """
+    block_number_response = plan.wait(
+        recipe=get_block_recipe(node_id, LATEST_BLOCK_NUMBER_GENERIC),
```

Comment:
Hah, yes I have to pad the numbers to be able to compare them as strings. Because it is true 5 < 3000, BUT, "0000005" > "0003000" does work. That's why I have the padding function

Reply:
@gbouv sweet, wonder if the pad is a quick hack we can code into the backend as well 😅  assuming of course that post padding the lexiographical order doesn't change for non number strings!

Reply:
Yeah, the proper fix here would be to inspect the objects being compared, and if they are numbers apply number comparison instead. But the issue is upstream here I believe, basically all runtime values are strings because of string replacement. So, that's probably what we should change first. Maybe when we tackle the project about FutureValues, we can incorporate this